message/messaging_service.cc

```diff
@@ -284,6 +284,26 @@ bool messaging_service::is_same_rack(inet_address addr) const {
     return topo.get_rack(addr) == topo.get_rack();
 }
 
+future<> messaging_service::ban_host(locator::host_id id) {
+    return container().invoke_on_all([id] (messaging_service& ms) -> future<> {
+        if (ms._banned_hosts.contains(id) || ms.is_shutting_down()) {
+            co_return;
+        }
+
+        ms._banned_hosts.insert(id);
+        auto [start, end] = ms._host_connections.equal_range(id);
+        for (auto it = start; it != end; ++it) {
+            auto& conn_ref = it->second;
+            conn_ref.server.abort_connection(conn_ref.conn_id);
+        }
+        ms._host_connections.erase(start, end);
```

Comment:
> What does ban_host() guarantee?

Only what the comment says:
```
// Drops all connections from the given host and prevents further communication from it to happen.
```

Perhaps I should add to the comment that any currently running handlers for this host will not be interrupted so no ambiguity is left. Will follow-up.

We cannot guarantee anything for work that was started concurrently, we can only make guarantees for future events (i.e. no further handlers will be called for RPCs coming from that node).

Reply:
I really meant, what guarantees should ban_host() make.

If calls can continue executing, then a left node can still make changes to the cluster, no?

Reply:
When you have two concurrent operations, you cannot guarantee anything about their order of execution. They can happen one after the other or they can interleave. This in particular applies when we consider the operation of banning a node happening concurrently with some RPC coming from that node. It's just theoretically impossible to give any guarantees here. By the time you call ban_host() the RPC handler might have already finished or it might be at the very end of its execution.

The left node can make some changes to the cluster but that's OK because the decision to make this change was made before it was left (banned). The important thing is that any decision made by it after it left the cluster won't affect the cluster.

Reply:
If ban_host() returns after all handlers have completed, it can guarantee that once it's complete, no operations from that host will take place.

Reply:
We'd have to introduce a bunch of complexity to be able to wait for all handlers that come from a particular hosts to finish (a gate per host or something). But I don't think we need this for anything. If a node made a call when it was still a member it's fine to finish the call. We want to prevent calls that it makes after it was removed, because then it could be using outdated information etc. 

If we find a use case for this we can always extend the mechanism with those gates in a follow-up, but for now I don't see any value in it.

Reply:
> If ban_host() returns after all handlers have completed, it can guarantee that once it's complete, no operations from that host will take place.

I had the same argument on the patch that introduced `abort_connection()` in the first place https://github.com/scylladb/seastar/pull/1663. Kamil convinced me that for this use case it is fine. You merged the patch, so I assume it was OK with you as well. Later @xemul introduced `server::shutdown()` https://github.com/scylladb/seastar/pull/1671 with the same semantics where I had the same argument and eventually was convinced that it is fine for the intended use case. You merged the patch.

Just the background...

Reply:
@avikivity are the answers satisfactionary? If yes - could you merge?

Reply:
> We'd have to introduce a bunch of complexity to be able to wait for all handlers that come from a particular hosts to finish (a gate per host or something). But I don't think we need this for anything. If a node made a call when it was still a member it's fine to finish the call. We want to prevent calls that it makes after it was removed, because then it could be using outdated information etc.
> 
> If we find a use case for this we can always extend the mechanism with those gates in a follow-up, but for now I don't see any value in it.

So really, we don't need to abort the connections, just prevent new ones.

I don't propose changing anything, just trying to understand. It seems to me that abort without waiting is a halfway measure and either banning new connections, or aborting+waiting would be more self-consistent solutions.

Reply:
> So really, we don't need to abort the connections, just prevent new ones.

If we don't abort the connections, then we would happily keep accepting new RPCs on these existing connections. This is because the "is host ID banned" check appears only once per connection - at the beginning when it's established and we handle the CLIENT_ID verb.

Reply:
@avikivity any chance for getting this merged?

Reply:
@avikivity I answered your last question. Please respond if you have further doubts or need further explanations.